CARDS-2703: Add selector to include Survey Event form link

acrowthe opened the pull request:
- Add `.formToSurveyLinks` processor which adds the uuid of a Survey Events form to the export of a form which contains a `belongsToForm` link
- Update QuestionnaireToCsvProcessor:
  - Check for arguments in the export path of the following format: `.csvIncludeFields:<value>=<label>` or `.csvIncludeFields:<value>="<label>"`
  - If present, export the `value` property that is on the form's json export base level into the `label` and/or raw `value` column
  - Adjust `created` and `lastModified` export process to use some of the same code to avoid duplication
- Update prems configuration to implement these changed for exports

Testing Instructions:
- Start in prems
- Create a Visit Information Form with Clinic = UHN Inpatient and Time = Yesterday
- Put some data into the generated Inpatient form and note it's path
- Use /bin/browser to update the Inpatient form:
  - Add `SUBMITTED` to the status flags
- Note the `jcr:uuid` of the Survey Events form
- Go to /system/console/configMgr and update `io.uhndata.cards.export.ExportConfig~UHN-Labeled-Forms`:
  - Change Data Storage parameters to `savePath=.`
- Go to http://localhost:8080/Subjects.export?config=UHN-Labeled-Forms
- Check the folder in which you launched CARDS: There should be multiple CSVs there, one for each prems patient form
  - Open the OAIP form, verify that the `@survey` column is present and filled out with the @name of the generated Survey Events form

acrowthe committed:
CARDS-2703: Add selector to include Survey Event form link
Add JsonProcessor that includes the uuid of Survey Event Forms when linked by belongsToSurvey links

acrowthe committed:
CARDS-2703: Add selector to include Survey Event form link
- Update QuestionnaireToCsvProcessor to add support for configurable properties
- Update prems configuration to include the new @survey property in the csv exports

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -376,7 +376,7 @@
       "retrieverParameters": [
         "questionnaire=/Questionnaires/Survey events",
-        "selectors=.dataFilter:statusNot=INCOMPLETE.labels"
+        "selectors=.dataFilter:statusNot=INCOMPLETE.labels.formToSurveyLinks"
```
We don't need this one.

marta- commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -147,6 +133,66 @@ private String processQuestionnaire(final JsonObject questionnaire, final Resour
+                if (endIndex < 0) {
+                    endIndex = resolutionPathInfo.length();
+                }
+                int divider = resolutionPathInfo.indexOf("=", startIndex);
+
+                String value = resolutionPathInfo.substring(startIndex + INCLUDE_FIELDS.length(),
+                    (divider < endIndex && divider > 0) ? divider : endIndex);
+                String label = (divider < endIndex && divider > 0)
+                    ? resolutionPathInfo.substring(divider + 1, endIndex)
+                    : value;
```
Not a big deal, but would something like this be easier to read?

```
String[] pieces = resolutionPathInfo.substring(startIndex + INCLUDE_FIELDS.length(), endIndex)
    .split("=", 2);
String value = pieces[0];
String label = pieces.length == 2 ? pieces[1] : pieces[0];

acrowthe committed:
CARDS-2703: Add selector to include Survey Event form link
Improve configuration and code cleanup

acrowthe committed:
CARDS-2703: Add selector to include Survey Event form link
Fix exported survey identifier to @name instead of jcr:uuid to match survey events export

veronikaslc commented on prems-resources/backend/pom.xml:
```diff
@@ -62,10 +62,20 @@
+    <dependency>
+      <groupId>javax.json</groupId>
+      <artifactId>javax.json-api</artifactId>
+      <version>1.1.4</version>
```
no need to specify version here

veronikaslc commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/serialize/FormToSurveyLinkProcessor.java:
```diff
@@ -0,0 +1,105 @@
+                return false;
+            }
+        }).findAny().orElse(null);
+
```
we can remove extra empty line here

veronikaslc commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/serialize/FormToSurveyLinkProcessor.java:
```diff
@@ -0,0 +1,105 @@
+        final Function<Node, JsonValue> serializeNode)
+    {
+        try {
+            if (node.isNodeType("cards:Form")) {
```
```suggestion
            if (node.isNodeType(FormUtils.FORM_NODETYPE)) {
```

veronikaslc commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/serialize/FormToSurveyLinkProcessor.java:
```diff
@@ -0,0 +1,105 @@
+        final Function<Node, JsonValue> serializeNode)
+    {
+        try {
+            if (child.isNodeType("cards:Links")) {
```
```suggestion
            if (child.isNodeType(LinkUtils.LINKS_NODETYPE)) {
```

acrowthe committed:
CARDS-2703: Add selector to include Survey Event form link
Code cleanup

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/serialize/FormToSurveyLinkProcessor.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public JsonValue processChild(final Node node, final Node child, final JsonValue input,
```
I don't think this method belongs here. We already have the `.nolinks` processor. Each processor should be focused on doing one thing, and this one should just add the `@survey` property.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/serialize/FormToSurveyLinkProcessor.java:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    private Link getBelongsToSurveyLink(Node linksNode)
```
Why not use the new `linkUtils.getLinksOfType(node, "belongsToSurvey")`?

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -399,7 +399,7 @@
         "questionnaire=/Questionnaires/Visit information",
         "questionnaire=/Questionnaires/Survey events",
-        "selectors=.csvHeader:raw.-dereference.dataOption:formSelectors=-dereference\\.simple\\.deep\\.-identify\\.nolinks"
+        "selectors=.csvHeader:raw.-dereference.dataOption:formSelectors=-dereference\\.simple\\.deep\\.-identify\\.formToSurveyLinks"
```
Leave the `nolinks` processor in place.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/serialize/FormToSurveyLinkProcessor.java:
I'd put this in `patient-portal`, it has to do with Survey Events, which are generic.

acrowthe committed:
CARDS-2703: Add selector to include Survey Event form link
- Move FormToSurveyLinksProcessor to patient portal
- Remove link node stripping from FormToSurveyLinksProcessor
- Switch over to using linkUtils.getLinkOfType
